CHANGELOG.md

```diff
@@ -29,7 +29,7 @@ And an assortment of other improvements:
 - `Text[Char]` instance was added for Postgres `COPY FROM STDIN`. Thanks **Dermot Haughey**!
 - You can now create a `HikariTransactor` from a `HikariConfig`. Thanks **Yuriy Badalyantc**!
 - Thanks to **Harry Laoulakos** and **Daan Hoogenboezem** for documentation updates!
-- `Fragment` combinators now parenthesize arguments so associativity in Scala is reflected in generated SQL. Thanks **Katrix**!
+- `Fragment` combinators now return parenthesized expressions so associativity in Scala is reflected in generated SQL. `and` and `or` & co. return an `Option[Fragment]`, `None` in case the input list is empty. This ensures that a valid SQL is always produced, i.e. no `... WHERE true AND ( ) `. Thanks **Katrix**!
 
```

Comment:
What's the argument against returning `true` and `false` for empty `and` and `or`, respectively?

Reply:
1. I'm not sure how well is the emptiness established (or could be) in the first place (let alone equality)
2. even if we had a reliable `isEmpty` method, it doesn't help me very well with composing/nesting `and`s or `or`s -- we actually, `andOpt`s and `orOpt`s to be precise
3 let's take a specific example:

```
val list1 = List()
val list2 = List(fr1, fr2)
...
andOpt(or(list1: _*), or(list2: _*))
```
this will generate exactly what I want
```
((fr1 OR fr2))
```
so not `() AND ((fr1) OR (fr2))` which is not a valid SQL

Reply:
Emptiness is well-defined for sequences and equality isn't relevant. Maybe I misunderstand.
 
My suggestion is that `and()` return `true` and `or()` return false, consistent with the boolean monoids, thus your `and(or(), or(a, b))` would return `(false and (a or b))`.

Reply:
> Emptiness is well-defined for sequences

I was talking about emptiness/equality of `Fragment`s, but only because I misunderstood you original suggestion _"returning true and false for empty and..."_, sorry :)
I now get what you mean!

> `and(or(), or(a, b))` would return `(false and (a or b))`
which is just `false`, btw.

as a user of these helper functions, I would consider that quite a surprising and undesirable behaviour -- opposite of what I want to achieve.
wouldn't also you as a newcomer be surprised by this?

I intuitively expect `and(or(), or(a, b))` to be essentially equivalent to `or(a, b)`

---

to use the language of monoid, the empty `or` in an `and` shouldn't be replaced by a neutral element for `or`, but by a neutral element of the enclosing function. which is in our case `and` and so the empty `or()` should be replaced by `true`

`X(Y(a, b), Y(c, d, e), Y())` `=>` `X(Y(a, b), Y(c, d, e), neutralX)`